docs/man/any2dance.1
.TH any2dance 1 "June 26th, 2003"
.SH NAME
any2dance \- convert any pyDDR-supported format to the .dance format
.SH SYNOPSIS
\fBany2dance [\-\-delete] \fIfilename.ext\fR | \fIdirectory\fR ...
.SH DESCRIPTION
This program converts step patterns in other formats readable by pyDDR,
to pyDDR's native .dance format. The dance format provides a superset of
most useful features in other file formats, and is the only format
pyDDR will always load correctly.
.PP
\fBany2dance\fR can be given an explicit filename to convert, or a directory
name. If it is given a directory, it will recurse down that directory
looking for files to convert.
.SH USAGE
.IP \-\-delete
If this option is given, the original file will be deleted after the
conversion is done. Be particularly careful about using this with
a directory argument.
.SH BUGS
DWI banner/background detection is not 100% accurate (blame DWI).
.SH AUTHORS
Brendan Becker <tgz@orotech.net> and Joe Wreschnig <piman@debian.org>.
.SH SEE ALSO
\fBpyddr\fR(6),
dance\-spec.txt included with pyDDR.


ChangeLog
pydance ChangeLog

legend:
+ new feature added or made an existing one better
- removed a feature, or something bad happened
* fixed a bug

----- - CURRENT
0.7.5 - That's the second largest ChangeLog I've ever seen!	[2003.XX.XX]
+ New options: "Wide" to add jumps, holds on/off.
+ 5 panel (Pump It Up) support (single, versus/unison, couple, and double).
 + Support for KSF, SM, and .dance files with 5 panel steps.
 - Actual PIU pads may or may not work; none were available for testing.
+ Versus, couple, and double 6 and 8 panel modes.
+ New game type selector, so you can do e.g. Double/Endless.
+ A document on the KSF file format has been added in docs/.
+ The .dance specification has been updated with new modes and corrections.
+ A new centered display for single player modes.
+ New theming code.
 + Brings back oldschool and bryan themes.
 + Autodetect what game modes a theme will work for.
+ New 32x32 arrow set for 8/9 panel modes, by xsdg.
* Support the Thrustmaster USB mats (probably), and X-Box controllers.
 + We suspect that pydance will run on GNU/Linux on an X-Box now.
* Correctly encode/decode KSF, SM, DWI, and .dance file metadata.
 * KSF requires the Python Koco codecs package.
* DWI parsing for braindead constructs like "2!(2".
* Arrows now always draw in order (earlier under later).
* Many crashes fixed:
 * Stops + holds works again.
 * Oni lifebar works again.
 * Obscure edge case in 2 player endless mode works correctly.
* Parsing for all files is now more informative and less error-prone.
- Assist mode is broken for non-4 panel modes, we need more sounds.
- any2dance removed, it won't work anymore with this version.
- In-game menu help removed, due to inaccuracies.

0.7.4 - In My Pants						[2003.07.28]
+ 6 panel ("solo") and 8 panel support.
+ Double (1 player, both sides) support.
 - 'totaljudgings' option no longer works (no one was using it anyway...)
* Song selector UI tweaks (faster animations, nicer folder opening).
* Respace the arrows to be closer together in single/versus.
* All 9 directions can always be done on the numeric pad for player 1.
 - azerty 2nd player start key changed to '+' on the numeric pad.
* "Escape" changed from left+start to start+select.
* Some more speed optimizations for slower processors.
* Some DWI parser fixes for bad files.
* 'usepsyco' option (on by default) to try and load the Psyco JIT compiler.
- 'oldschool' and 'bryan' were removed (temporarily, maybe).

0.7.3
* Skipped, because the Linux Game Tome called 0.7.2 0.7.3 by mistake.

0.7.2 -	Dilaudid, how I love thee                               [2003.07.14]
+ Boost, brake, left, right, mirror, shuffle, and random ("supershuffle").
+ "Suck" lifebar mode, you fail if you do too good.
+ "Drop" lifebar mode, goes down at the regular speed but never up.
+ Folders in the song selector (needs > 60 songs).
* Parser bug fixes (especially for SM files).
* pyDDR has been renamed to pydance.
	
0.7.1 - OUCH MY BLADDER                                         [2003.07.01]
+ Support for COUPLE mode.
 + And battle mode.
 - any2dance doesn't always convert COUPLE from DWI to .dance correctly.
+ New player options screen (hold down start in the song selector).
+ New game options screen (hold down select).
+ "Ready? Go!" screen before the song starts.
+ New "oldschool" theme similar to earlier mixes.
* Standard round of bug fixes and tweaks.

0.7.0 - Applesauce makes the apples go yonder                   [2003.06.04]
+ More optimizations from JZig (movies this time)
+ New file formats
  + Native DWI and SM support
  + ".dance" instead of ".step" is now pyDDR's official format.
  + 'any2dance.py' converter for files.
* Judge rewrite!
  * Proper speed change handling in Judge
  * Proper miss handling in Judge

0.6.5 - Boy, did I mess that up.				[2003.05.30]
+ Major graphics optimization patch (1.5x-2x increase in fps) from JZig.
* Two player mode works again.

0.6.4 - RTFM. RTFM. RTFM. That means Read the Fucking Manual.	[2003.05.29]
+ New step file parsers
  + Beginner and hardcore (smaniac/oni/challenge) difficulty.
+ New hold arrow display
- 'mixerclock' option removed (it's now the only timing method).
- 'transcolor' option removed in favor of new 'lyriccolor' syntax.
* Bug fixes, optimizations, whippings, etc.
  + Most notably, holds, bpm changes, and stops are working much nicer now.
    - But may require reconversion with dwi2step.

0.6.3 - Reasonable Starting Times!				[2003.05.17]
+ New song loading display and algorithm (at least 4-5 times faster).
+ New song selector based roughly on DDR 5th Mix.
  + Nonstop help and playlist display.
  + 'gratuitous' option to turn off the slow extras in the song selector.
+ Endless mode.
+ Configurable oni life meter.
+ New "centered" arrow scrolling style for gameplay.
+ Smaller download size by converting wavs to oggs.
  - Requires SDL_mixer 1.2.5.
+ The timer display is controllable via the 'fpsdisplay' option.
+ New theme ('blue') and a new life display for all themes.
* DDR Extreme (8th Mix) grading and scoring algorithm.
* Split off the ddrmat.o driver from the main package.
* The constant cleanup/speedup/refactor/bug fix process.

0.6.2 - No More Ugly-Ass Buttons				[2003.04.28]
- Fullscreen switching during the game doesn't work on MacOS X.
  - Use the configuration file instead, for now.
- The pN_keys option is gone in favor of 'keyboard', and the 'sortpersist'
  option was removed.
+ 'killsongonfail' was renamed to 'autofail' and made default.
+ New game menu display.
+ Better theme detection and configuration support.
+ The 'badknees' option will remove jumps.
+ A new README that actually documents all the options!
* Once again, numerous cleanups and bugs fixed.

0.6.1 - The second one						[2003.04.09]
+ Experimental install system
  + Currently only works on Windows (and doesn't do much) and UNIX.
  + Manual pages included with pyDDR now.
  + New directory layout
+ Major code cleanups and bug fixes.

any2dance.py
#!/usr/bin/env python
# Convert any 

import sys, os, getopt, fnmatch, time

pydance_path = sys.argv[0]
if os.name == "posix":
  sys.path.append(os.path.split(os.path.realpath(pydance_path))[0])
else: sys.path.append(os.path.split(os.path.abspath(pydance_path))[0])

VERSION = "0.2"
FORMATS = ('*.step', '*.dance', '*.dwi', '*.sm')

import fileparsers

def convert(arg, delete):
  if not os.path.exists(arg):
    sys.stderr.write("no such filename: " + arg + "\n")
  elif os.path.isdir(arg):
    for file in os.listdir(arg): convert(os.path.join(arg, file), delete)
  else:
    matched = False
    for format in FORMATS:
      if fnmatch.fnmatch(arg, format): matched = True
    if not matched: return

    cwd = os.getcwd()
    dir, name = os.path.split(arg)
    if dir == "": dir = "."
    os.chdir(dir)
    try:
      newname = name[:name.rindex(".")] + ".dance"
      print "Converting", name, "to", newname

      song = fileparsers.SongItem(name)
      dir = os.path.realpath(".")
      for k in ("banner", "background", "filename", "movie"):
        if song.info[k]: song.info[k] = song.info[k].replace(dir + os.sep, '')
      song.write(newname)

      if delete: os.unlink(name)
    except:
      sys.stderr.write("Error converting " + name + "\n")
    os.chdir(cwd)

def usage(exitcode = 0):
  print "any2dance", VERSION, "- convert other formats to .dance"
  print "Usage:", sys.argv[0], "oldfile.ext"
  sys.exit(exitcode)

def main():

  delete = False

  try: opts, args = getopt.getopt(sys.argv[1:], "vh",
                                  ["version", "help", "delete"])
  except getopt.GetoptError: usage(1)

  for o, a in opts:
    if o in ("--help", "-h"):
      usage()
    elif o in ("-v", "--version"):
      sys.stderr.write("any2dance version " + VERSION + "\n")
      sys.exit()
    elif o in ("--delete",):
      print "WARNING: Files will be deleted after conversion. You have 5 seconds to exit."
      time.sleep(5)
      delete = True

  if len(args) == 0: usage(2)

  for arg in args: convert(arg, delete)

if __name__ == '__main__': main()
